sstables/sstables_manager.cc

```diff
@@ -149,8 +149,40 @@ sstable_writer_config sstables_manager::configure_writer(sstring origin) const {
     return cfg;
 }
 
-void sstables_manager::update_total_memory_used_by_components(const sstable* sst) {
+void sstables_manager::update_total_memory_used_by_components_and_maybe_reclaim(const sstable* sst) {
     _total_memory_used_by_components += sst->total_components_memory_size();
+
+    if (_total_memory_used_by_components <= _max_allowed_memory_for_components) {
+        // total memory used is within limit; no need to reclaim.
+        return;
+    }
+
+    // Memory consumption has crossed threshold; Some of it needs to be reclaimed.
+    // Sort the _active list in descending order of total memory, so that we can
+    // start reclaiming from the largest sstable in terms of memory.
+    _active.sort([] (const sstable& sst1, const sstable& sst2) {
+        return sst1.total_components_memory_size() > sst2.total_components_memory_size();
+    });
+
```

Comment:
this is a potential source of reactor stall. I suggest you futurize this function and add a maybe_yield in the sort

a node can have thousands of sstables (1000÷0.16 = ~6k), so we can end up violating the task quota.

Reply:
We can use std::max_element and only drop the largest one.

I think we're guaranteed to succeed: since exactly one sstable caused us to go over the edge, then dropping that sstable would cause us to go below. Dropping the largest one would be even better. It's ok (though not great) to cycle over thousands of sstables.

quota / nr_sstables = 0.5ms / 10,000 = 50ns/iteration, that's 100 cycles/iteration, sufficient for max_element.

Reply:
also: you can store total components size inside the sstable itself, no need to recalculate it.

Reply:
maybe max_element is wrong, if we have a mix if reclaimable and non-reclaimable components.

Reply:
You're right. Maybe we can live with a partial sort and restrict each reclaim cycle to the nr of sorted elements.

Reply:
> maybe max_element is wrong, if we have a mix if reclaimable and non-reclaimable components.

which is why I suggested to rename total_components_memory_size to total_reclaimable_memory_size() which is how it's implemented today. If a sstable today has its filter evicted, that function returns zero.

Reply:
On a second thought, what Avi said above is correct. As we're checking the limit on every sstable addition, each reclaim cycle needs to purge exactly one filter in order to cause the limit to be respected. So max_element is enough.

Reply:
Ack. `std:max_element` was indeed sufficient and we needed to reclaim from only one sstable to get total usage under limit. I have updated the method to reflect this.